src/persist/src/operators/stream.rs

```diff
@@ -41,46 +50,215 @@ where
 {
     fn persist(
         &self,
-        token: (StreamWriteHandle<K, V>, StreamReadHandle<K, V>),
+        name: &str,
+        write: StreamWriteHandle<K, V>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    ) {
+        let operator_name = format!("persist({})", name);
+        let mut persist_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut input = persist_op.new_input(&self, Pipeline);
+
+        let (mut data_output, data_output_stream) = persist_op.new_output();
+        let (mut error_output, error_output_stream) = persist_op.new_output();
+
+        let mut buffer = Vec::new();
+        let mut write_futures = HashMap::new();
+        let mut input_frontier =
+            Antichain::from_elem(<G::Timestamp as timely::progress::Timestamp>::minimum());
+
+        persist_op.build(move |_capabilities| {
+            move |frontiers| {
+                let mut data_output = data_output.activate();
+                let mut error_output = error_output.activate();
+
+                // write out everything and forward, keeping the write futures
+                input.for_each(|cap, data| {
+                    data.swap(&mut buffer);
+
+                    let write_future = write.write(buffer.iter().as_ref());
+
+                    let mut session = data_output.session(&cap);
+                    session.give_vec(&mut buffer);
+
+                    let write_futures = &mut write_futures
+                        .entry(cap.retain_for_output(1))
+                        .or_insert_with(|| Vec::new());
+                    write_futures.push(write_future);
+                });
+
+                // Block on outstanding writes when the input frontier advances.
+                // This way, when the downstream frontier advances, we know that all writes that
+                // are before it are done.
+                let new_input_frontier = frontiers[0].frontier();
+                let progress =
+                    !PartialOrder::less_equal(&new_input_frontier, &input_frontier.borrow());
+
+                if progress {
+                    input_frontier.clear();
+                    input_frontier.extend(new_input_frontier.into_iter().cloned());
+                    if let Some(data_frontier) = input_frontier.elements().first() {
```

Comment:
I'm uncomfortable hard coding single element antichains where it's avoidable. I think this just wants to iterate the keys and keep them if `!input_frontier.less_equal(key)` no?